### PR TITLE
txn: check committed record before async commit fallback

### PR DESCRIPTION
**Pull request by sticnarf**
Signed-off-by: Yilin Chen <sticnarf@gmail.com>

### What is changed and how it works?

Issue Number: Close https://github.com/tikv/tikv/issues/12615

What's Changed:

```commit-message
If a retried prewrite encounters CommitTsTooLarge, it may cause correctness
issues (#12615). So, like other errors, this commit adds a check for committed
records to make prewrite idempotent.
```

### Related changes

- Need to cherry-pick to the release branch

### Check List 

Tests 

- Unit test

Side effects

- Performance regression

It increases cost on `CommitTsTooLarge`. But this error should not happen very frequently in normal production.

### Release note 

```release-note
Fix possible duplicate commit record in async-commit pessimistic transactions.
```

### CODE REVIEW

**Review by cfzjywxk — APPROVED** [review]

File: src/storage/txn/commands/prewrite.rs [hunk 1]
```diff
@@ -528,6 +528,13 @@ impl<K: PrewriteKind> Prewriter<K> {
                     return check_committed_record_on_err(prewrite_result, txn, reader, &key);
                 }
                 Err(MvccError(box MvccErrorInner::CommitTsTooLarge { .. })) | Ok((..)) => {
+                    // The prewrite might be a retry and the record may have been committed.
+                    // So, we need to prevent the fallback to avoid duplicate commits.
+                    if let Ok(res) =
+                        check_committed_record_on_err(prewrite_result, txn, reader, &key)
```
Comment (youjiali1995) [inline src/storage/txn/commands/prewrite.rs:534]: L530 matches `Ok((..))`. Is it expected? BTW, I forget why it matches `OK((..))` here.

Comment (sticnarf) [reply src/storage/txn/commands/prewrite.rs:547]: Sadly, this definitely causes some problems... If it needs min_commit_ts but min_commit_ts is zero, the lock
has been prewritten and has fallen back from async commit or 1PC.

**Comment by sticnarf** [comment]
@youjiali1995 Do you mind taking a look again?

File: src/storage/txn/commands/prewrite.rs (hunk 1 above)
Comment (sticnarf) [reply src/storage/txn/commands/prewrite.rs:547]: Yes.

Comment (sticnarf) [reply src/storage/txn/commands/prewrite.rs:547]: > I'm a bit confused. It seems `CommitTsTooLarge` is consumed inside and this error would not be returned no matter if it's a retried request or not?

`CommitTsTooLarge` is always consumed and not returned to the client...

**Pull request merged**